### DISCUSSION TITLE
Correct Way to Use XComs (Backend)

### DISCUSSION THREAD

**Discussion by hbc-acai** [apache/airflow#35519] [Q&A]
Hello community, 

I am new to XCom. I am reading the [official documentation](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/xcoms.html) and some other community posts, and seeing conflicting information. 

The official document says: 
"An XCom is identified by a key (essentially its name), as well as the task_id and dag_id it came from. They can have any (serializable) value, but they are only designed for small amounts of data; do not use them to pass around large values, like dataframes."

First question - if XCom variable is only identified by the tuple `(name, task_id, dag_id)`, does it mean it is not aware of the execution_date? If I run trigger the dags as of two different execution dates, will XCom variable override each other in a race condition? 

Second condition - if I use a proper XCom backend like S3, is it fine to pass around large values? If the official document is correct, then the ETL example in this [post](https://tegardp.medium.com/the-6-step-etl-process-using-airflow-with-example-and-exercise-db46715a61f0) would be wrong

Thanks,

**Comment by Joffreybvn** [comment]
Hello,
Sharing my experience on XCom backends:

**First question**: XComs won't override themself. An Xcom is identified by its [*dag_id*, *task_id*, *run_id* and *map_index*](https://github.com/apache/airflow/blob/1b059c57d6d57d198463e5388138bee8a08591b1/airflow/models/xcom.py#L393-L397). If you implement your S3 backend like described by [this astronomer tutorial](https://docs.astronomer.io/learn/xcom-backend-tutorial), each file will also have a random uuid. Thus, if you have multiple Airflow instances with the same S3 bucket as backend, running the same dag at the same time in each instance would not create any collision.

**Second question**: Yes, you can. But beware of download / upload cost (in time, and $). I had S3 disk as backend, until I switched to a simple Volume (on a Kubernetes cluster). Some dags that were passing lots of data via XComs got their execution time reduced by a big 30%. Depending on where your Airflow is living, S3 may not be the definitive solution.
Also, on Airflow side, some features are not ready for big data, like the [RenderingTaskInstanceFields](https://github.com/apache/airflow/issues/28199). Quickly summarized: Using big chunk of data in templated fields will result in having this data stored in the airflow database. First time I did it, my Postgres did not liked it. You can disable this behavior.
And finally, some operators tend to throw a lot of logs (like the SQLExecuteQuery which prints all the data twice when performing INSERT statements). Airflow webserver will crash if you try to see the logs of a task were all your data was printed out. Fortunately, upoming 2.8 release of Airflow will allow to better configure logging, and avoiding this case.

**Reply by hbc-acai** [reply]
Thanks. It is super helpful. 
1. Looks like the document is little bit outdated. 
2. I will follow that RenderingTaskInstanceFields issue. 
I can also see the returned value is log. It can be a very lone line. 
[2023-11-09, 15:10:19 EST] {process_utils.py:186} INFO - Output: 
[2023-11-09, 15:10:57 EST] {python.py:194} INFO - Done. Returned value was: gASVKQYAAAAAAACMGW ****
